lib/class-wp-rest-url-details-controller.php

```diff
@@ -208,13 +214,93 @@ private function get_remote_url( $url ) {
 	 * @return string the title tag contents (maybe empty).
 	 */
 	private function get_title( $html ) {
-		preg_match( '|<title>([^<]*?)</title>|is', $html, $match_title );
+		preg_match( '|<\s*title[^>]*>(.*?)<\s*/\s*title>|is', $html, $match_title );
 
-		$title = isset( $match_title[1] ) ? trim( $match_title[1] ) : '';
+		$title = isset( $match_title[1] ) && is_string( $match_title[1] ) ? trim( $match_title[1] ) : '';
 
 		return $title;
 	}
 
+	/**
+	 * Parses the site icon from the provided HTML
+	 *
+	 * @param string $html the HTML from the remote website at URL.
+	 * @param string $url the target website URL.
+	 * @return string the icon URI (maybe empty).
+	 */
+	private function get_icon( $html, $url ) {
+		preg_match( '|<link.*?rel="\s*[shortcut]+(?:\s+[icon]+)*\s*".*?href="(.*?)".*?\/?>|is', $html, $matches );
+
+		$icon = isset( $matches[1] ) && is_string( $matches[1] ) ? trim( $matches[1] ) : '';
+
+		// Attempt to convert relative URLs to absolute.
+		if ( ! empty( $icon ) ) {
+			$parsed = parse_url( $icon );
+			if ( empty( $parsed['host'] ) ) {
+				$icon = $url . $icon;
+			}
+		}
+
+		return $icon;
+	}
+
+	/**
+	 * Parses the meta description from the provided HTML.
+	 *
+	 * @param string $html the HTML from the remote website at URL.
+	 * @return string the meta description contents (maybe empty).
+	 */
+	private function get_description( $html ) {
```

Comment:
@hellofromtonya How do you feel about the writing to a `tmp` file here simply to make use of `get_meta_tags`?

Reply:
It's a clever way to use native PHP functionality. However, I'm concerned about working with the filesystem: server permissions, leaving behind artifacts, performance, etc. 

For this specific parsing method, it's too much for parsing just the `<meta name="description" ... >` element. Why? The meta element has a clear well-formed signature (via the `name` attribute). Here, regex will be more understandable and performant without the potential filesystem gotchas that some sites will experience.

Side note:
This type of code, i.e. get all meta elements, could be something for pre-parsing the `head` _before_ invoking each specific parser (i.e. before getting the title, icon, description, image, etc.). However, it would need consideration for performance, memory handling, cognitive complexity, etc. I'm noting it here for thought when we start working on tuning the performance of getting the metadata.

Reply:
Unfortunately, it seems to only return a very specific subset of `meta` tags. At least when I tested.

Let's switch to regex here then.